File: pkg/ikuai_api4/tag_name.go
package ikuai_api4

import (
	"regexp"
	"strconv"
	"strings"

	"ikuai-bypass/pkg/ikuai_common"
	"ikuai-bypass/pkg/logger"
)

var tagNameSanitizer = regexp.MustCompile(`[^\p{Han}A-Za-z0-9]+`)

// 包级 logger 用于打印截断警告 / Package-level logger for truncation warnings
var tagLogger = logger.NewLogger("TAG:名称处理")

func stripKnownPrefix(raw string) string {
	raw = strings.TrimSpace(raw)
	raw = strings.TrimPrefix(raw, ikuai_common.COMMENT_IKUAI_BYPASS+"_")
	raw = strings.TrimPrefix(raw, ikuai_common.NAME_PREFIX_IKB)
	return strings.TrimSpace(raw)
}

// sanitizeTagName 移除特殊符号，仅保留中文、英文和数字。
func sanitizeTagName(raw string) string {
	return tagNameSanitizer.ReplaceAllString(stripKnownPrefix(raw), "")
}

// buildTagName 统一构建规则/分组名字，固定前缀 IKB。
func buildTagName(raw string) string {
	token := sanitizeTagName(raw)
	if token == "" {
		return ikuai_common.NAME_PREFIX_IKB
	}
	return ikuai_common.NAME_PREFIX_IKB + token
}

const maxTagNameLength = 15 // 爱快 4.0.101 对 tagname 的长度限制 / iKuai 4.0.101 tagname length limit

func buildIndexedTagName(raw string, index int) string {
	suffix := strconv.Itoa(index + 1)
	baseName := buildTagName(raw)
	originalName := baseName + suffix
	
	// 如果总长度超过限制，截断 baseName
	// Truncate baseName if total length exceeds limit
	maxBaseLen := maxTagNameLength - len(suffix)
	if len(baseName) > maxBaseLen {
		truncatedName := baseName[:maxBaseLen] + suffix
		tagLogger.Warn("TRUNCATE:名称截断", "Tag name truncated due to iKuai 15-char limit: '%s' -> '%s' (original tag: '%s')", originalName, truncatedName, raw)
		return truncatedName
	}
	
	return originalName
}

func buildTagNameCandidates(raw string) []string {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return nil
	}
	candidateSet := map[string]struct{}{}
	add := func(v string) {
		v = strings.TrimSpace(v)
		if v == "" {
			return
		}
		candidateSet[v] = struct{}{}
	}

	// 支持逗号分隔
	for _, part := range strings.Split(raw, ",") {
		token := sanitizeTagName(part)
		if token != "" {
			add(ikuai_common.NAME_PREFIX_IKB + token)
		}
	}

	res := make([]string, 0, len(candidateSet))
	for k := range candidateSet {
		res = append(res, k)
	}
	return res
}

// matchTagNameFilter 增加“名字 + 旧配置(comment)”的兼容匹配。
func matchTagNameFilter(filterTagName, currentName, legacyComment string) bool {
	if strings.TrimSpace(filterTagName) == "" {
		return true
	}
	// 强制校验 currentName 是否以 IKB 开头，除非它是旧版备注匹配
	isManaged := strings.HasPrefix(currentName, ikuai_common.NAME_PREFIX_IKB)

	for _, c := range buildTagNameCandidates(filterTagName) {
		// 如果名字以候选词开头（处理序号），且是受管规则
		if isManaged && strings.HasPrefix(currentName, c) {
			return true
		}
		// 兼容性：包含匹配（慎用，主要用于备注）
		if legacyComment != "" && strings.Contains(legacyComment, c) {
			return true
		}
	}
	return false
}

File: crates/core/src/ikuai/tag_name.rs
use md5;
use once_cell::sync::Lazy;
use regex::Regex;

use super::types::{COMMENT_IKUAI_BYPASS, NAME_PREFIX_IKB};

static TAG_SANITIZER: Lazy<Result<Regex, regex::Error>> =
    Lazy::new(|| Regex::new(r"[^\p{Han}A-Za-z0-9]+"));

const MAX_TAG_NAME_LENGTH: usize = 15;

const IP_GROUP_RAND_MARKER: &str = "R";

fn strip_known_prefix(raw: &str) -> String {
    let mut s = raw.trim().to_string();
    let legacy = format!("{}{}", COMMENT_IKUAI_BYPASS, "_");
    if s.starts_with(&legacy) {
        s = s.trim_start_matches(&legacy).trim().to_string();
    }
    if s.starts_with(NAME_PREFIX_IKB) {
        s = s.trim_start_matches(NAME_PREFIX_IKB).trim().to_string();
    }
    s
}

pub fn sanitize_tag_name(raw: &str) -> String {
    let base = strip_known_prefix(raw);
    match TAG_SANITIZER.as_ref() {
        Ok(re) => re.replace_all(&base, "").to_string(),
        Err(_) => base,
    }
}

pub fn build_tag_name(raw: &str) -> String {
    let token = sanitize_tag_name(raw);
    if token.is_empty() {
        return NAME_PREFIX_IKB.to_string();
    }
    truncate_utf8_by_bytes(
        &format!("{}{}", NAME_PREFIX_IKB, token),
        MAX_TAG_NAME_LENGTH,
    )
}

pub fn build_indexed_tag_name(raw: &str, index: i64) -> String {
    let suffix = (index + 1).to_string();
    let base = build_tag_name(raw);
    let max_base_len = MAX_TAG_NAME_LENGTH.saturating_sub(suffix.len());
    let base = truncate_utf8_by_bytes(&base, max_base_len);
    format!("{}{}", base, suffix)
}

pub fn build_indexed_tag_name_with_affix(raw: &str, affix: &str, index: i64) -> String {
    let suffix = (index + 1).to_string();
    let base = build_tag_name(raw);
    let max_base_len = MAX_TAG_NAME_LENGTH.saturating_sub(affix.len() + suffix.len());
    let base = truncate_utf8_by_bytes(&base, max_base_len);
    format!("{}{}{}", base, affix, suffix)
}

fn hash_token_letters(raw: &str, len: usize) -> String {
    let len = len.clamp(1, 6);
    let digest = md5::compute(raw.trim().as_bytes());
    let mut out = String::with_capacity(len);
    for i in 0..len {
        let b = digest.0.get(i).copied().unwrap_or(0);
        let ch = (b % 26) as u8 + b'A';
        // Avoid using marker letter to make parsing easier.
        // 避免使用 marker 字母，便于解析。
        let ch = if ch == b'R' { b'S' } else { ch };
        out.push(ch as char);
    }
    out
}

/// Build indexed tag name for ip-group / ipv6-group with a deterministic random-like suffix.
/// 为 IP 分组（IPv4/IPv6）构建带“随机后缀”的分片名称（确定性 hash token）。
pub fn build_indexed_ip_group_tag_name(raw: &str, index: i64) -> String {
    let token = hash_token_letters(raw, 2);
    let affix = format!("{}{}", IP_GROUP_RAND_MARKER, token);
    build_indexed_tag_name_with_affix(raw, &affix, index)
}

fn truncate_utf8_by_bytes(s: &str, max_bytes: usize) -> String {
    if s.len() <= max_bytes {
        return s.to_string();
    }
    let mut end = 0usize;
    for (i, ch) in s.char_indices() {
        let next = i + ch.len_utf8();
        if next > max_bytes {
            break;
        }
        end = next;
    }
    s[..end].to_string()
}

fn build_tag_name_candidates(raw: &str) -> Vec<String> {
    let raw = raw.trim();
    if raw.is_empty() {
        return Vec::new();
    }
    let mut set = std::collections::BTreeSet::new();
    for part in raw.split(',') {
        let token = sanitize_tag_name(part);
        if !token.is_empty() {
            set.insert(format!("{}{}", NAME_PREFIX_IKB, token));
        }
    }
    set.into_iter().collect()
}

pub fn match_tag_name_filter(filter_tag: &str, current_name: &str, legacy_comment: &str) -> bool {
    if filter_tag.trim().is_empty() {
        return true;
    }
    let is_managed = current_name.trim().starts_with(NAME_PREFIX_IKB);
    for c in build_tag_name_candidates(filter_tag) {
        let name = current_name.trim();
        if is_managed && name.starts_with(&c) {
            return true;
        }

        // Handle auto-truncated names caused by iKuai tagname length limit.
        // 兼容爱快 tagname 长度限制导致的自动截断（避免无法识别/无法原地更新）。
        if is_managed {
            let base0 = name.trim_end_matches(|ch: char| ch.is_ascii_digit());
            let base = strip_ip_group_rand_affix(base0);
            if !base.is_empty() && c.starts_with(base) {
                return true;
            }
        }
        if !legacy_comment.is_empty() && legacy_comment.contains(&c) {
            return true;
        }
    }
    false
}

fn strip_ip_group_rand_affix(s: &str) -> &str {
    // Our deterministic random-like suffix format is: "R" + 2 uppercase letters.
    // 确定性随机后缀格式："R" + 2 位大写字母。
    let b = s.as_bytes();
    if b.len() < 3 {
        return s;
    }
    let n = b.len();
    if b[n - 3] != b'R' {
        return s;
    }
    let a = b[n - 2];
    let c = b[n - 1];
    if !(a.is_ascii_uppercase() && c.is_ascii_uppercase()) {
        return s;
    }
    // ASCII suffix => safe to slice by bytes.
    &s[..n - 3]
}

#[cfg(test)]
mod tests {
    use super::build_tag_name;

    #[test]
    fn build_tag_name_truncates_to_ikuai_limit() {
        let tag = build_tag_name("SafeChunkRoute");
        assert_eq!(tag, "IKBSafeChunkRou");
        assert_eq!(tag.len(), 15);
    }
}
